File: cdm-sources/cdm/observable/asset/calculatedrate/functions/IndexValueObservationMultiple.java
package cdm.observable.asset.calculatedrate.functions;

import cdm.observable.asset.InterestRateIndex;
import com.google.inject.ImplementedBy;
import com.rosetta.model.lib.functions.RosettaFunction;
import com.rosetta.model.lib.mapper.MapperC;
import com.rosetta.model.lib.mapper.MapperS;
import com.rosetta.model.lib.records.Date;
import java.math.BigDecimal;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import javax.inject.Inject;


@ImplementedBy(IndexValueObservationMultiple.IndexValueObservationMultipleDefault.class)
public abstract class IndexValueObservationMultiple implements RosettaFunction {
	
	// RosettaFunction dependencies
	//
	@Inject protected IndexValueObservation indexValueObservation;

	/**
	* @param observationDate 
	* @param interestRateIndex 
	* @return observedValues 
	*/
	public List<BigDecimal> evaluate(List<Date> observationDate, InterestRateIndex interestRateIndex) {
		List<BigDecimal> observedValues = doEvaluate(observationDate, interestRateIndex);
		
		return observedValues;
	}

	protected abstract List<BigDecimal> doEvaluate(List<Date> observationDate, InterestRateIndex interestRateIndex);

	public static class IndexValueObservationMultipleDefault extends IndexValueObservationMultiple {
		@Override
		protected List<BigDecimal> doEvaluate(List<Date> observationDate, InterestRateIndex interestRateIndex) {
			if (observationDate == null) {
				observationDate = Collections.emptyList();
			}
			List<BigDecimal> observedValues = new ArrayList<>();
			return assignOutput(observedValues, observationDate, interestRateIndex);
		}
		
		protected List<BigDecimal> assignOutput(List<BigDecimal> observedValues, List<Date> observationDate, InterestRateIndex interestRateIndex) {
			observedValues.addAll(MapperC.<Date>of(observationDate)
				.mapItem(item -> MapperS.of(indexValueObservation.evaluate(item.get(), interestRateIndex))).getMulti());
			
			return observedValues;
		}
	}
}

File: cdm-sources/src/cdm/observable/asset/calculatedrate/functions/IndexValueObservationMultiple.py
# pylint: disable=line-too-long, invalid-name, missing-function-docstring, missing-module-docstring, superfluous-parens
# pylint: disable=wrong-import-position, unused-import, unused-wildcard-import, wildcard-import, wrong-import-order, missing-class-docstring
from __future__ import annotations
import sys
import datetime
import inspect
from decimal import Decimal
from rosetta.runtime.utils import *
from rosetta.runtime.func_proxy import replaceable, create_module_attr_guardian
from cdm.observable.asset.InterestRateIndex import InterestRateIndex
from cdm.observable.asset.calculatedrate.functions.IndexValueObservation import IndexValueObservation

__all__ = ['IndexValueObservationMultiple']


@replaceable
def IndexValueObservationMultiple(observationDate: list[datetime.date] | None, interestRateIndex: InterestRateIndex) -> Decimal:
    """
    Retrieve the values of the supplied index on the specified observation dates.
    
    Parameters 
    ----------
    observationDate : date
    
    interestRateIndex : InterestRateIndex
    
    Returns
    -------
    observedValues : number
    
    """
    self = inspect.currentframe()
    
    
    observedValues = map(lambda item: IndexValueObservation(item, rosetta_resolve_attr(self, "interestRateIndex")), rosetta_resolve_attr(self, "observationDate"))
    
    
    return observedValues

sys.modules[__name__].__class__ = create_module_attr_guardian(sys.modules[__name__].__class__)
